File: 프로그래머스/3/161988. 연속 펄스 부분 수열의 합/연속 펄스 부분 수열의 합.js
function solution(sequence) {
    
    var len = sequence.length;
    
    const arr = Array.from({length : len}, () => Array(2).fill(0));
    
    arr[0][0] = sequence[0];
    arr[0][1] = -sequence[0];
    
    for(let i=1; i<len; i++){
        arr[i][0] = Math.max(arr[i-1][1], 0) + sequence[i];
        arr[i][1] = Math.max(arr[i-1][0], 0) - sequence[i];
    }
    
    var answer = 0;
    
    for(let ar of arr){
        answer = Math.max(answer, Math.max(ar[0], ar[1]));
    }
    
    return answer;
}

File: 프로그래머스/3/161988. 연속 펄스 부분 수열의 합/연속 펄스 부분 수열의 합.java
class Solution {
    public long solution(int[] sequence) {
        
        int len = sequence.length;
        
        long dp[][] = new long[len][2];
        
        dp[0][0] = sequence[0];
        dp[0][1] = -sequence[0];
        
        for(int i=1; i<len; i++){
            dp[i][0] = Math.max(dp[i-1][1], 0) + sequence[i];
            dp[i][1] = Math.max(dp[i-1][0], 0) - sequence[i];
        }
        
        long answer = 0;
        
        for(long[] arr : dp){
            answer = Math.max(answer, Math.max(arr[0], arr[1]));
        }
        
        return answer;
    }
}